components/VM_Arm/src/modules/virtio_con.c

```diff
@@ -20,63 +19,133 @@
+static virtio_con_t *virtio_con = NULL;
 extern vmm_pci_space_t *pci;
 extern vmm_io_port_list_t *io_ports;
 
-struct {
+/* 4088 because the serial_shmem_t struct has to be 0x1000 bytes big */
+#define BUFSIZE (0x1000 - 2 * sizeof(uint32_t))
+
+/* This struct occupies exactly 1 page and represents the data in the shmem region */
+typedef struct serial_shmem {
     uint32_t head;
     uint32_t tail;
     char buf[BUFSIZE];
-} extern volatile *batch_buf;
+} serial_shmem_t;
+compile_time_assert(serial_shmem_1k_size, sizeof(serial_shmem_t) == 0x1000);
 
-static int handle_serial_console(vm_t *vmm, void *cookie UNUSED)
-{
-    struct {
-        uint32_t head;
-        uint32_t tail;
-        char buf[BUFSIZE];
-    } volatile *buffer = serial_getchar_buf;
+/*
+ * Represents the virtqueues used in a given link between
+ * two VMs.
+ */
+typedef struct serial_conn {
+    void (*notify)(void);
+    volatile serial_shmem_t *recv_buf;
+    volatile serial_shmem_t *send_buf;
+} serial_conn_t;
+static serial_conn_t connections[NUM_PORTS];
 
+/* This is a temporary buffer used to stage data before completing the virtio RX */
+static char temp_rx_buf[BUFSIZE];
+
+void virtio_console_putchar(int port, virtio_emul_t *con, char *buf, int len) WEAK;
+static void virtio_con_notify_recv(int port)
+{
+    volatile serial_shmem_t *buffer = connections[port].recv_buf;
     int count = 0;
     while (buffer->head != buffer->tail) {
-        txbuf[count % BUFSIZE] = buffer->buf[buffer->head];
+        temp_rx_buf[count] = buffer->buf[buffer->head];
         buffer->head = (buffer->head + 1) % BUFSIZE;
-        count += 1 % BUFSIZE;
+        count++;
+    }
+
+    assert(count <= BUFSIZE);
+    virtio_console_putchar(port, virtio_con->emul, temp_rx_buf, count);
+}
+
+static int handle_serial_console(vm_t *vmm, void *cookie UNUSED)
+{
+    /* Poll each ring buffer to see if data was added */
+    for (int i = 0; i < NUM_PORTS; i++) {
+        if (connections[i].recv_buf->head != connections[i].recv_buf->tail) {
+            virtio_con_notify_recv(i);
+        }
     }
-    virtio_console_putchar(virtio_con->emul, txbuf, count);
     return 0;
 }
 
-static void emulate_console_putchar(char c)
+static void tx_virtcon_forward(int port, char c)
 {
 #if CONFIG_NUM_DOMAINS > 1 && CONFIG_PRINTING
     seL4_DebugPutChar(c);
 #else
+    if (port >= NUM_PORTS) {
+        return;
+    }
+
+    volatile serial_shmem_t *batch_buf = connections[port].send_buf;
     batch_buf->buf[batch_buf->tail] = c;
     batch_buf->tail = (batch_buf->tail + 1) % BUFSIZE;
-    if (c == '\n' || batch_buf->tail + 1 == batch_buf->head) {
-        batch_batch();
+
+    /* If newline or staging area full, it's time to send */
+    if (c == '\n' || (batch_buf->tail + 1) % BUFSIZE == batch_buf->head) {
+        connections[port].notify();
     }
 #endif
 }
 
+/* regions shared by the SerialServer and the VMM */
+extern serial_shmem_t volatile *batch_buf;
+extern serial_shmem_t volatile *serial_getchar_buf;
+
```

Comment:
It's confusing to have the same-named extern variables here and local in tx_virtcon_forward()

Reply:
extern variables ideally are in header files.

Reply:
It seems to be a "conventional" way to deal with camkes-generated symbols (putting them right above where they're needed).

Reply:
Actually not, the proper header templates seem to be missing sometimes.